### ISSUE TITLE
Wrong Proton version used; failure to detect official Proton versions?

### ISSUE THREAD

**Issue by michaelneverwins** [sonic2kk/steamtinkerlaunch#123]
### Problem
When I try to run Giana Sisters: Twisted Dreams (app ID 223220) by putting `stl %command%` in the game's launch options, the `MP10Setup.exe` steps work but then the game fails to run because `proton-5.13-1b` is used, even though STL seems to be trying to configure the game for `proton-5.0-9`.

#### Details
When this occurs, I get the following notifications from STL:
* Using 'proton-5.13-1b' because Proton override is not enabled
* Proton version 'proton-5.0-9' required, but '' is set - switching
* Starting game 'Giana Sisters Twisted Dreams (223220)' using proton-5.13-1b

The log at `~/.config/stl/logs/223220.log` is a bit more clear about the probable cause. I can paste the whole log if needed, but here is an excerpt of what looks like the most relevant part:
```
Sat Oct 31 10:19:29 EDT 2020 INFO - launchSteamGame - GAMESTARTCMD is '/home/michael/.steam/steam/steamapps/common/Proton 5.13/proton waitforexitandrun /home/michael/.steam/steam/steamapps/common/Giana Sisters Twisted Dreams/launcher/GSLauncher.exe'
Sat Oct 31 10:19:29 EDT 2020 INFO - launchSteamGame - USESLR is enabled, so adding RUNSLR '/home/michael/.steam/steam/steamapps/common/SteamLinuxRuntime_soldier/_v2-entry-point --deploy=soldier --suite=soldier --verb=waitforexitandrun --' to the actual start command
Sat Oct 31 10:19:29 EDT 2020 INFO - wantProton - Proton version 'proton-5.0-9' required, but '' is set
Sat Oct 31 10:19:29 EDT 2020 INFO - switchProton - Switching used Proton and its Variables to version 'proton-5.0-9'
Sat Oct 31 10:19:29 EDT 2020 INFO - switchProton - Tried to switch to Proton 'proton-5.0-9', but it is not available
Sat Oct 31 10:19:29 EDT 2020 INFO - switchProton - AUTOPULLPROTON is enabled, so trying to download and enable 'proton-5.0-9'
Sat Oct 31 10:19:29 EDT 2020 INFO - createDLProtList - generating list of online available custom Proton builds
Sat Oct 31 10:19:31 EDT 2020 SKIP - switchProton - no download URL found for requested 'proton-5.0-9' - skipping
Sat Oct 31 10:19:31 EDT 2020 INFO - launchSteamGame - Proton Override allowed, so checking if it needs updated
Sat Oct 31 10:19:31 EDT 2020 INFO - launchSteamGame - no current Proton found
Sat Oct 31 10:19:31 EDT 2020 INFO - launchSteamGame - automatically selecting newest official one
Sat Oct 31 10:19:31 EDT 2020 INFO - setNOP - Selecting newest official Proton available
Sat Oct 31 10:19:31 EDT 2020 SKIP - setNOP - haven't found anything
Sat Oct 31 10:19:31 EDT 2020 INFO - launchSteamGame - command line proton '/home/michael/.steam/steam/steamapps/common/Proton 5.13/proton' is identical to 'RUNPROTON' (path for '') - nothing to change
Sat Oct 31 10:19:31 EDT 2020 INFO - launchSteamGame - using ORIGINAL game start command: /home/michael/.steam/steam/steamapps/common/Proton 5.13/proton waitforexitandrun /home/michael/.steam/steam/steamapps/common/Giana Sisters Twisted Dreams/launcher/GSLauncher.exe
Sat Oct 31 10:19:32 EDT 2020 INFO - setProtonPathVersion Setting the Proton Version for '/home/michael/.steam/steam/steamapps/common/Proton 5.13/proton' to 'proton-5.13-1b'
Sat Oct 31 10:19:32 EDT 2020 INFO - launchSteamGame - starting 'Giana Sisters Twisted Dreams (223220)' with Proton: 'proton-5.13-1b'
Sat Oct 31 10:19:32 EDT 2020 INFO - launchSteamGame - full start command is '/home/michael/.steam/steam/steamapps/common/SteamLinuxRuntime_soldier/_v2-entry-point --deploy=soldier --suite=soldier --verb=waitforexitandrun -- /home/michael/.steam/steam/steamapps/common/Proton 5.13/proton waitforexitandrun /home/michael/.steam/steam/steamapps/common/Giana Sisters Twisted Dreams/launcher/GSLauncher.exe'
```
When I run `stl settings`, only the custom Proton versions I've installed manually (Proton 5.9-GE-5-ST and Proton 5.9-GE-8-ST) are appearing in the Proton Version drop-down menu. Steam has installed official Proton versions to `~/.steam/steam/steamapps/common`, e.g. `~/.steam/steam/steamapps/common/Proton 5.0` which is presumably the 5.0-9 version which I can select in Steam, but STL doesn't seem to detect it.

I'm using SteamTinkerLaunch v2.6.0, with YAD 7.3, on Linux Mint 19.3.

### Workarounds
I can get the game to run if I go to the game's properties in Steam and explicitly force the use of Proton 5.0-9 there, while keeping `stl %command%` in the launch options field. However, I assume this shouldn't be necessary.

**Comment by michaelneverwins** [comment]
I thought maybe I had done something wrong the first time I used STL, so I tried to get a fresh start by deleting both `~/.config/stl/` and `/dev/shm/stl`, and ran `stl settings` again. I think I succeeded in resetting all of the relevant configuration, but `stl settings` printed some errors (and I think I remember seeing these the first time I ever ran STL as well):
```
find: ‘/usr/share/steam/compatibilitytools.d’: No such file or directory
find: ‘/home/michael/.steam/root/steamapps/common’: No such file or directory
find: ‘/steamapps/common’: No such file or directory
```
I'm not sure if these errors are relevant, and maybe some are expected if STL tries to account for multiple possible system configurations by checking locations that may or may not exist on any given system, but I'm suspicious of the second and third errors. The third looks like some environment variable wasn't set, unless some users actually have a `/steamapps` at the root of the file system. As for the second error: The directory `/home/michael/.steam/root/steamapps/common` does not exist, but `/home/michael/.steam/root` is a link to `/home/michael/.steam` which contains a `steam/steamapps/common`, so `/home/michael/.steam/root/steam/steamapps/common` does exist. (Is this not the normal directory structure for Steam on Linux?)

As an experiment, I made a slight change to my copy of the `stl` script:
```diff
diff --git a/stl b/stl
index d3250e4..79f1ace 100755
--- a/stl
+++ b/stl
@@ -727,7 +727,7 @@ function getAvailableProtonVersions {
                        if [ -f "$PROTBIN" ]; then
                                fillProtonCSV
                        fi
-               done <<< "$(find "$HSR/$SAC" -maxdepth 1 -maxdepth 1 -type d -name "Proton*")"  
+               done <<< "$(find "$HSR/steam/$SAC" -maxdepth 1 -maxdepth 1 -type d -name "Proton*")"
 
                # official proton versions installed via Steam in additional SteamLibrary Paths
                if [ -f "$HSRCC" ]; then
```
This gets `stl settings` to populate the Proton Versions drop-down list with the official versions. However, this isn't to say I actually know what I'm doing; that one-line hack doesn't really fix anything in terms of running the game. If I use this modified script to run Giana Sisters, with `stl %command%` in the launch options and no Proton version forced via Steam, STL does try to run the game with `proton-5.0-9` but then it crashes.

The same crash also occurs if I revert the script back to normal and instead create a link from `/home/michael/.steam/root/steamapps/` to `/home/michael/.steam/root/steam/steamapps/` so that the expected directory `/home/michael/.steam/root/steamapps/common` exists. Meanwhile, however, the workaround of forcing Proton 5.0-9 via the game's properties in Steam still works. So maybe, even if STL can find Proton 5.0-9, it's having trouble switching to it. Or maybe I'm just really good at breaking things.

**Comment by frostworx** [comment]
Looks like you really tried to fix the issue, so thank you first for a good bug report!
There were already several Giana and `~/.steam/root/steamapps/common` path issues in the past and I'm not sure how much they are related. 
(Currently) stl doesn't install official Steam Proton versions, so if it is really not installed, it has to installed manually (I'm not sure if stl should do this automatically, so it might be better to be simply more specific in the logs).
With one of the last recent changes I changed the `WANTPROTON` value in the Giana tweak to proton-[5.0-9](https://github.com/frostworx/steamtinkerlaunch/blob/master/tweaks/user/223220.conf#L13) which was `WANTPROTON="proton-5.0-10-rc4" before.`
**So if you have the "next-" braunch of Proton 5.0 installed, it is internally at version "proton-5.0-10-rc4" and therefore the WANTPROTON version "proton-5.0-9" is not found**.
This clearly requires a proper fix for 
"allow minor Proton version changes (untested) (fixed #101)"
introduced with the last version 2.6.0.

Independent of Giana it looks like the whole `~/.steam/root/steamapps/common`paths should be exported into user configurable variables _(IMHO the whole sym-linking is chaos on its own and even crazier if it changes from time to time or from distribution to distribution)_.

May I ask how you installed your yad 7.3 version? Might be interesting for  issue #124

**Comment by frostworx** [comment]
Last time playing with the strange paths I added this little check:
https://github.com/frostworx/steamtinkerlaunch/blob/master/stl#L157
Looks like we just need the same for `steamapps/common` to get usable paths on debian flavours. I'll merge everything into one initial check function as summed up here #126

**Comment by frostworx** [comment]
edit: As nearly every steam path was _(**is** until #126 is fixes)_ broken with current configuration it is obviously the main reason for no valid proton being found. Thanks again for a very good and important report!

**Comment by michaelneverwins** [comment]
> Independent of Giana it looks like the whole `~/.steam/root/steamapps/common` paths should be exported into user configurable variables _(IMHO the whole sym-linking is chaos on its own and even crazier if it changes from time to time or from distribution to distribution)_.

If it changes from time to time then I wonder if my current configuration differs at all from that of a freshly installed Steam. I've had it installed for quite a while now. I don't think I manually did anything weird to its directory structure, though.

> May I ask how you installed your yad 7.3 version? Might be interesting for issue #124

I had some false starts when it came to installing YAD 7.3 (mostly because I had failed to read the dependencies carefully enough), but once I got it right, the steps were basically as follows:
```bash
sudo apt install automake autoconf intltool libgtk-3-dev  # (YAD build dependencies per README)
mkdir -p ~/repositories && cd ~/repositories  # (or wherever you like to keep your cloned repos)
git clone https://github.com/v1cont/yad.git yad-dialog-code
cd yad-dialog-code
autoreconf -ivf && intltoolize
./configure && make && sudo make install
gtk-update-icon-cache
```
Notes:
* Cloning into `~/repositories` is just my own personal preference. The location doesn't really matter.
* The `sudo` on `make install` was not mentioned in YAD's README, but I added it anyway because otherwise it failed to install to my `root`-owned `/usr/local/bin` directory.
* The YAD repo's README indicates that the last command, `gtk-update-icon-cache`, needs to be executed manually after installation. I actually forgot to run that command until later, and STL still worked. However, I'm including the command anyway because the YAD documentation says it should be included.
* **EDIT:** It looks like my install steps for YAD are almost identical to the steps previously described in a comment on another issue (https://github.com/frostworx/steamtinkerlaunch/issues/98#issuecomment-706558925). I could have saved myself a bit of time if I had seen that before! I think that comment's author and I are both essentially following the instructions from [YAD's README file](https://github.com/v1cont/yad/blob/master/README.md), but in my opinion the README doesn't spell it out quite as clearly (and, of course, it also doesn't list the exact names of all dependencies as they appear in Linux Mint's package manager).

**Comment by frostworx** [comment]
> If it changes from time to time then I wonder if my current configuration differs at all from that of a freshly installed Steam. I've had it installed for quite a while now. I don't think I manually did anything weird to its directory structure, though.

Hehe, I have exactly the same problem :)
I don't know if it changed from time to time, it was just a wild guess because it would be a half-sane explanation of all those symlinks` (my ~/.steam/root/steam/ is a real directory btw with just a `cached` and a `games` dir and a `legacydepotdata.vdf` file)`

I'll try to fix #126 as fast as possible _(I hope today)_
Thanks for the yad info, so it looks like chances are pretty high that #124 really is an appimage problem!
Will mention it in the sticky issue

**Comment by frostworx** [comment]
Just wrote this little path test script here:
https://github.com/frostworx/steamtinkerlaunch/issues/126#issuecomment-721285773
Would be nice if you could check if all paths are resolved on your system

**Comment by michaelneverwins** [comment]
> Would be nice if you could check if all paths are resolved on your system

I think they were all resolved. The script didn't seem to be giving me any errors. See my reply on the other issue: https://github.com/frostworx/steamtinkerlaunch/issues/126#issuecomment-721333321

**Comment by frostworx** [comment]
nice, thanks for testing :+1:

**Comment by frostworx** [comment]
*fixed* in https://github.com/frostworx/steamtinkerlaunch/releases/tag/v2.7.0
Please reopen if there are still issues

**Comment by michaelneverwins** [comment]
> _fixed_ in https://github.com/frostworx/steamtinkerlaunch/releases/tag/v2.7.0
> Please reopen if there are still issues

Thanks @frostworx; I will re-test when I get a chance (maybe tonight or later this week).

**Comment by frostworx** [comment]
Thanks as well @michaelneverwins! Take your time, it doesn't run away :)

**Comment by michaelneverwins** [comment]
I tried re-testing with the latest changes, and I'm still having some kind of problem. However, it wasn't a very thorough test. I might be re-opening this issue this weekend, but I won't do so until I have something useful to share. (And, of course, I'd also like to rule out the possibility that I'm just doing something wrong.)

Maybe I should test against the v2.7.0 release linked above, instead of pulling all the latest changes. (My installed `~/.local/bin/stl` is actually just a symbolic link to the `stl` in my clone of this repository, so I updated it by doing a `git fetch && git pull` and got more changes than I was expecting.) Then again, Steam has apparently replaced my Proton 5.0-9 with a Proton 5.0-10 since I last tested this, so I guess it's already too late to test v2.7.0 under *exactly* the same conditions under which I tested v2.6.0.

**Comment by frostworx** [comment]
Always possible that a new bug appears, but does stl even find all bundled tweaks in your setup? Maybe a `stl update git` is already enough, which also pulls in the recently changed tweaks for giana

**Comment by michaelneverwins** [comment]
> does stl even find all bundled tweaks in your setup?

Actually I think the script downloads them into `~/.config/stl/`; it doesn't detect the ones that already exist in my clone of the repository, if that's what you're asking, so my setup isn't really any better than if I had just downloaded the `stl` script by itself, but I didn't know that until after I tried using it.

**Comment by frostworx** [comment]
True, stl doesn't search for configs in `pwd`
What you could do instead is symlinking `stl` from the `stl update git` to your ~/.local/bin/stl.

**Comment by michaelneverwins** [comment]
I tested again, and whatever problem I'm having does not seem to have gone away. I'd re-open the issue but I don't think I have permission to do so.

Sorry for the massive info dump below, but if you're willing to look into this, I don't want to waste your time by leaving out important information.

#### STL v2.7.0

I checked out the v2.7.0 commit and made sure my installed link was pointing to it,
```bash
cd ~/repositories/steamtinkerlaunch
git fetch
git pull
git checkout v2.7.0
ln -sfn ~/repositories/steamtinkerlaunch/stl ~/.local/bin/stl
cd
```
cleared out the settings,
```bash
rm -rf /dev/shm/stl ~/.config/stl
```
and ran `stl settings` again. (Actually, I had to run it twice, because it fails due to a missing language file the first time I run it after deleting `~/.config/stl/`, but this happened with v2.6.0 as well.)

It still prints a single `find` error to the terminal,
```
find: ‘/steamapps/common’: No such file or directory
```
and the Proton Version drop-down menu in the settings window shows only my custom Proton versions, Proton-5.9-GE-5-ST and Proton-5.9-GE-8-ST, as well as a couple of blank entries (which was also the case before, with v2.6.0, but I forgot to mention it).

![stl-1](https://user-images.githubusercontent.com/73746653/98447311-cd63a980-20f1-11eb-8d7e-ce48f340601c.png)

Then I cleared out Steam's Wine prefix for Giana Sisters: Twisted Dreams,
```bash
rm -rf .steam/steam/steamapps/compatdata/223220
```
and ran the game via Steam with `stl %command%` in the launch options. The `MP10Setup.exe` tweak was applied as expected, but then the game crashed, likely because the wrong Proton version was used. This time, it ended up using Proton-5.9-GE-8-ST, according to the notifications sent by STL:
* Using 'proton-5.13-1b' because Proton override is not enabled
* Downloading 'MP10Setup.exe' from ...
* Extracting 'MP10Setup.exe' into ...
* Proton version 'proton-5.0-10b' required, but '' is set - switching
* Minor Proton version mismatch - using 'XXX' instead of requested 'YYY'
* Starting game 'Giana Sisters Twisted Dreams (223220)' using "Proton-5.9-GE-8-ST"
* 'Giana Sisters Twisted Dreams (223220)' Stopped. Exiting SteamTinkerLaunch

The contents of `~/.config/stl/logs/223220.log` might be more helpful. As before, I've pasted what appears to be the relevant part below:
```
Sat Nov  7 12:08:25 EST 2020 INFO - launchSteamGame - GAMESTARTCMD is '/home/michael/.steam/steam/steamapps/common/Proton 5.13/proton waitforexitandrun /home/michael/.steam/steam/steamapps/common/Giana Sisters Twisted Dreams/launcher/GSLauncher.exe'
Sat Nov  7 12:08:25 EST 2020 INFO - launchSteamGame - USESLR is enabled, so adding RUNSLR '/home/michael/.steam/steam/steamapps/common/SteamLinuxRuntime_soldier/_v2-entry-point --deploy=soldier --suite=soldier --verb=waitforexitandrun --' to the actual start command
Sat Nov  7 12:08:25 EST 2020 INFO - wantProton - Proton version 'proton-5.0-10b' required, but '' is set
Sat Nov  7 12:08:26 EST 2020 INFO - wmp10toPFX - removing two wine dlls which come uppercase with 'MP10Setup.exe'
Sat Nov  7 12:08:26 EST 2020 INFO - switchProton - Switching used Proton and its Variables to version ''
Sat Nov  7 12:08:26 EST 2020 INFO - wmp10toPFX - 'MP10Setup.exe' dlls have been successfully installed into '/home/michael/.steam/steam/steamapps/compatdata/223220/pfx/drive_c/windows/syswow64'
Sat Nov  7 12:08:26 EST 2020 INFO - switchProton - requested Proton version '' found under '/home/michael/.steam/root/compatibilitytools.d/Proton-5.9-GE-8-ST/proton'
Sat Nov  7 12:08:26 EST 2020 INFO - setNewProtVars - Setting new Proton Variables based on '/home/michael/.steam/root/compatibilitytools.d/Proton-5.9-GE-8-ST/proton'
Sat Nov  7 12:08:26 EST 2020 INFO - setNewProtVars - set the wine binary for proton in path '/home/michael/.steam/root/compatibilitytools.d/Proton-5.9-GE-8-ST/proton' to '/home/michael/.steam/root/compatibilitytools.d/Proton-5.9-GE-8-ST/dist/bin/wine'
Sat Nov  7 12:08:26 EST 2020 INFO - setProtonPathVersion Setting the Proton Version for '/home/michael/.steam/root/compatibilitytools.d/Proton-5.9-GE-8-ST/proton' to '"Proton-5.9-GE-8-ST"'
Sat Nov  7 12:08:26 EST 2020 INFO - fixProtonVersionMismatch - minor version mismatch - using '' instead of requested 'proton-5.0-10b'
Sat Nov  7 12:08:26 EST 2020 INFO - launchSteamGame - Proton Override allowed, so checking if it needs updated
Sat Nov  7 12:08:26 EST 2020 INFO - launchSteamGame - ORIGINAL game start command is: /home/michael/.steam/steam/steamapps/common/Proton 5.13/proton waitforexitandrun /home/michael/.steam/steam/steamapps/common/Giana Sisters Twisted Dreams/launcher/GSLauncher.exe
Sat Nov  7 12:08:26 EST 2020 INFO - launchSteamGame - overriding 'proton-5.13-1b' (='/home/michael/.steam/steam/steamapps/common/Proton 5.13/proton') from command line with '"Proton-5.9-GE-8-ST"' (='/home/michael/.steam/root/compatibilitytools.d/Proton-5.9-GE-8-ST/proton')
Sat Nov  7 12:08:26 EST 2020 INFO - launchSteamGame - game was started with 'stl' as Launch option, so replacing '/home/michael/.steam/steam/steamapps/common/Proton 5.13/proton' with '/home/michael/.steam/root/compatibilitytools.d/Proton-5.9-GE-8-ST/proton' in '/home/michael/.steam/steam/steamapps/common/Proton 5.13/proton waitforexitandrun /home/michael/.steam/steam/steamapps/common/Giana Sisters Twisted Dreams/launcher/GSLauncher.exe'
Sat Nov  7 12:08:26 EST 2020 INFO - launchSteamGame - new launch command is '/home/michael/.steam/root/compatibilitytools.d/Proton-5.9-GE-8-ST/proton waitforexitandrun /home/michael/.steam/steam/steamapps/common/Giana Sisters Twisted Dreams/launcher/GSLauncher.exe'
Sat Nov  7 12:08:26 EST 2020 INFO - launchSteamGame - UPDATED game start command is: /home/michael/.steam/root/compatibilitytools.d/Proton-5.9-GE-8-ST/proton waitforexitandrun /home/michael/.steam/steam/steamapps/common/Giana Sisters Twisted Dreams/launcher/GSLauncher.exe
Sat Nov  7 12:08:26 EST 2020 INFO - setProtonPathVersion Setting the Proton Version for '/home/michael/.steam/root/compatibilitytools.d/Proton-5.9-GE-8-ST/proton' to '"Proton-5.9-GE-8-ST"'
Sat Nov  7 12:08:26 EST 2020 INFO - launchSteamGame - starting 'Giana Sisters Twisted Dreams (223220)' with Proton: '"Proton-5.9-GE-8-ST"'
Sat Nov  7 12:08:26 EST 2020 INFO - launchSteamGame - full start command is '/home/michael/.steam/steam/steamapps/common/SteamLinuxRuntime_soldier/_v2-entry-point --deploy=soldier --suite=soldier --verb=waitforexitandrun -- /home/michael/.steam/root/compatibilitytools.d/Proton-5.9-GE-8-ST/proton waitforexitandrun /home/michael/.steam/steam/steamapps/common/Giana Sisters Twisted Dreams/launcher/GSLauncher.exe'
```

I didn't force any specific Proton version in the game's properties, but it's probably worth noting that I have my Proton version set to 5.13-1 in the global Steam settings. I assume this should be the default, because it's the latest, but for some reason, this setting stuck to Proton 5.0-9 when the new Proton 5.13-1 was first released, and I subsequently moved it to Proton 5.13-1 manually. So if you're trying to reproduce these results, you might want to make you've selected Proton 5.13-1 in the global settings.

![stl-2](https://user-images.githubusercontent.com/73746653/98447527-71018980-20f3-11eb-84fd-9ebd55f01b07.png)

Anyway, it looks like there are multiple things going on here.
* It doesn't seem to be detecting official Proton versions, and `''` is set by default.
* It's incorrectly using Proton-5.9-GE-8-ST now, whereas originally it was incorrectly using Proton 5.13-1.

#### STL latest

I repeated the test above using the latest version of STL,
```bash
cd ~/repositories/steamtinkerlaunch
git checkout master
git fetch
git pull
ln -sfn ~/repositories/steamtinkerlaunch/stl ~/.local/bin/stl
cd
rm -rf /dev/shm/stl ~/.config/stl
stl settings
```
and the results were only slightly different.

Running `stl settings` didn't show any `find` error, and instead I got some different terminal output,
```
Sat Nov  7 12:27:20 EST 2020 SKIP - checkProtonBump - need Proton version as argument 1
```
but the Proton Version drop-down menu looked the same as with v2.6.0 and v2.7.0.

When I tried running the game again, the notifications were the same, as far as I could tell. Proton-5.9-GE-8-ST was used again. The `launchSteamGame` logging, timestamps aside, also appears to be identical.

**Comment by michaelneverwins** [comment]
By the way, my installed Proton versions are in the following locations:
```
/home/michael/.steam/steam/steamapps/common/Proton 5.13
/home/michael/.steam/steam/steamapps/common/Proton 5.0
/home/michael/.steam/steam/steamapps/common/Proton 3.7
/home/michael/.steam/steam/steamapps/common/Proton 4.2
/home/michael/.steam/steam/steamapps/common/Proton 4.11
/home/michael/.steam/steam/steamapps/common/Proton 3.16
/home/michael/.steam/compatibilitytools.d/Proton-5.9-GE-8-ST
/home/michael/.steam/compatibilitytools.d/Proton-5.9-GE-5-ST
```
I found these using
```bash
find ~/.steam -type d -name "Proton*"
```
but I don't think you would want to do something so inefficient in the script.

Using the `~/.steam/root -> ~/.steam` link, the locations would be
```
/home/michael/.steam/root/steam/steamapps/common/Proton 5.13
/home/michael/.steam/root/steam/steamapps/common/Proton 5.0
/home/michael/.steam/root/steam/steamapps/common/Proton 3.7
/home/michael/.steam/root/steam/steamapps/common/Proton 4.2
/home/michael/.steam/root/steam/steamapps/common/Proton 4.11
/home/michael/.steam/root/steam/steamapps/common/Proton 3.16
/home/michael/.steam/root/compatibilitytools.d/Proton-5.9-GE-8-ST
/home/michael/.steam/root/compatibilitytools.d/Proton-5.9-GE-5-ST
```

**Comment by michaelneverwins** [comment]
I decided that perhaps I was too quick to assume only the `launchSteamGame` logging was relevant, so I took another look at `~/.config/stl/logs/223220.log` (which meant generating a new one because I had cleared out my `~/.config/stl/` again).

In particular, I wanted to see the `setSteamPath` logging, but it doesn't seem to show any problems:
```
Sat Nov  7 14:17:55 EST 2020 SKIP - setSteamPath - 'SROOT' already defined as '/home/michael/.steam/root/'
Sat Nov  7 14:17:55 EST 2020 SKIP - setSteamPath - 'SUSDA' already defined as '/home/michael/.steam/steam/userdata'
Sat Nov  7 14:17:55 EST 2020 SKIP - setSteamPath - 'DEFSTEAMAPPS' already defined as '/home/michael/.steam/steam/steamapps'
Sat Nov  7 14:17:55 EST 2020 SKIP - setSteamPath - 'DEFSTEAMAPPSCOMMON' already defined as '/home/michael/.steam/steam/steamapps/common'
Sat Nov  7 14:17:55 EST 2020 SKIP - setSteamPath - 'CFGVDF' already defined as '/home/michael/.steam/steam/config/config.vdf'
Sat Nov  7 14:17:56 EST 2020 SKIP - setSteamPath - 'STEAMCOMPATOOLS' already defined as '/home/michael/.steam/root/compatibilitytools.d'
```
That `DEFSTEAMAPPSCOMMON` path is where my official Proton versions are located.

So I looked at the code, and I *might have* found the problem:
```bash
                if [ -d "$DEFSTEAMAPPSCOMMON" ]; then
                        while read -r protondir; do
                                PROTBIN="$protondir/proton"
                                if [ -f "$PROTBIN" ]; then
                                        fillProtonCSV
                                fi
                        done <<< "$(find "$DEFSTEAMAPPS" -maxdepth 1 -maxdepth 1 -type d -name "Proton*")"
                else
                        writelog "SKIP" "${FUNCNAME[0]} - directory DEFSTEAMAPPS '$DEFSTEAMAPPSCOMMON' not found - this should not happen! - skipping"
                fi
```
You're checking for the existence of the `$DEFSTEAMAPPSCOMMON` directory but then searching inside of `$DEFSTEAMAPPS` with maximum depth of 1, so it wouldn't find `"$DEFSTEAMAPPS/common/Proton 5.0"` etc. So I made a small change:
```diff
diff --git a/stl b/stl
index a6c9c7b..625f403 100755
--- a/stl
+++ b/stl
@@ -753,7 +753,7 @@ function getAvailableProtonVersions {
                                if [ -f "$PROTBIN" ]; then
                                        fillProtonCSV
                                fi
-                       done <<< "$(find "$DEFSTEAMAPPS" -maxdepth 1 -maxdepth 1 -type d -name "Proton*")"      
+                       done <<< "$(find "$DEFSTEAMAPPSCOMMON" -maxdepth 1 -maxdepth 1 -type d -name "Proton*")"
                else
                        writelog "SKIP" "${FUNCNAME[0]} - directory DEFSTEAMAPPS '$DEFSTEAMAPPSCOMMON' not found - this should not happen! - skipping"
                fi
```
Now when I run `stl settings` after clearing out `~/.config/stl/` and `/dev/shm/stl/`, the Proton Versions list is populated with official versions as well as the custom ones, although there's still a blank space for some reason:

![stl-3](https://user-images.githubusercontent.com/73746653/98451504-1c223b00-2114-11eb-9f9a-c8d637267229.png)

Unfortunately, this only gets me as far as the hack described earlier &mdash; see https://github.com/frostworx/steamtinkerlaunch/issues/123#issuecomment-719970197 &mdash; because, when I launch the game, it detects the correct Proton version and says it's switching, but then the game crashes. Here's the `launchSteamGame` logging for that attempt:
```
Sat Nov  7 14:57:48 EST 2020 INFO - launchSteamGame - GAMESTARTCMD is '/home/michael/.steam/steam/steamapps/common/Proton 5.13/proton waitforexitandrun /home/michael/.steam/steam/steamapps/common/Giana Sisters Twisted Dreams/launcher/GSLauncher.exe'
Sat Nov  7 14:57:48 EST 2020 INFO - launchSteamGame - USESLR is enabled, so adding RUNSLR '/home/michael/.steam/steam/steamapps/common/SteamLinuxRuntime_soldier/_v2-entry-point --deploy=soldier --suite=soldier --verb=waitforexitandrun --' to the actual start command
Sat Nov  7 14:57:48 EST 2020 INFO - wantProton - Proton version 'proton-5.0-10b' required, but '' is set
Sat Nov  7 14:57:48 EST 2020 INFO - wantProton - Proton version 'proton-5.0-10b' does exist - switching
Sat Nov  7 14:57:48 EST 2020 INFO - switchProton - Switching used Proton and its Variables to version 'proton-5.0-10b'
Sat Nov  7 14:57:48 EST 2020 INFO - switchProton - requested Proton version 'proton-5.0-10b' found under '/home/michael/.steam/steam/steamapps/common/Proton 5.0/proton'
Sat Nov  7 14:57:49 EST 2020 INFO - setNewProtVars - Setting new Proton Variables based on '/home/michael/.steam/steam/steamapps/common/Proton 5.0/proton'
Sat Nov  7 14:57:49 EST 2020 INFO - setNewProtVars - set the wine binary for proton in path '/home/michael/.steam/steam/steamapps/common/Proton 5.0/proton' to '/home/michael/.steam/steam/steamapps/common/Proton 5.0/dist/bin/wine'
Sat Nov  7 14:57:49 EST 2020 INFO - setProtonPathVersion Setting the Proton Version for '/home/michael/.steam/steam/steamapps/common/Proton 5.0/proton' to 'proton-5.0-10b'
Sat Nov  7 14:57:49 EST 2020 INFO - launchSteamGame - Proton Override allowed, so checking if it needs updated
Sat Nov  7 14:57:49 EST 2020 INFO - launchSteamGame - ORIGINAL game start command is: /home/michael/.steam/steam/steamapps/common/Proton 5.13/proton waitforexitandrun /home/michael/.steam/steam/steamapps/common/Giana Sisters Twisted Dreams/launcher/GSLauncher.exe
Sat Nov  7 14:57:49 EST 2020 INFO - launchSteamGame - overriding 'proton-5.13-1b' (='/home/michael/.steam/steam/steamapps/common/Proton 5.13/proton') from command line with 'proton-5.0-10b' (='/home/michael/.steam/steam/steamapps/common/Proton 5.0/proton')
Sat Nov  7 14:57:49 EST 2020 INFO - launchSteamGame - game was started with 'stl' as Launch option, so replacing '/home/michael/.steam/steam/steamapps/common/Proton 5.13/proton' with '/home/michael/.steam/steam/steamapps/common/Proton 5.0/proton' in '/home/michael/.steam/steam/steamapps/common/Proton 5.13/proton waitforexitandrun /home/michael/.steam/steam/steamapps/common/Giana Sisters Twisted Dreams/launcher/GSLauncher.exe'
Sat Nov  7 14:57:49 EST 2020 INFO - launchSteamGame - new launch command is '/home/michael/.steam/steam/steamapps/common/Proton 5.0/proton waitforexitandrun /home/michael/.steam/steam/steamapps/common/Giana Sisters Twisted Dreams/launcher/GSLauncher.exe'
Sat Nov  7 14:57:49 EST 2020 INFO - launchSteamGame - UPDATED game start command is: /home/michael/.steam/steam/steamapps/common/Proton 5.0/proton waitforexitandrun /home/michael/.steam/steam/steamapps/common/Giana Sisters Twisted Dreams/launcher/GSLauncher.exe
Sat Nov  7 14:57:49 EST 2020 INFO - setProtonPathVersion Setting the Proton Version for '/home/michael/.steam/steam/steamapps/common/Proton 5.0/proton' to 'proton-5.0-10b'
Sat Nov  7 14:57:49 EST 2020 INFO - launchSteamGame - starting 'Giana Sisters Twisted Dreams (223220)' with Proton: 'proton-5.0-10b'
Sat Nov  7 14:57:49 EST 2020 INFO - launchSteamGame - full start command is '/home/michael/.steam/steam/steamapps/common/SteamLinuxRuntime_soldier/_v2-entry-point --deploy=soldier --suite=soldier --verb=waitforexitandrun -- /home/michael/.steam/steam/steamapps/common/Proton 5.0/proton waitforexitandrun /home/michael/.steam/steam/steamapps/common/Giana Sisters Twisted Dreams/launcher/GSLauncher.exe'
```

So if you agree with my change to the script, then I think the problem is no longer with finding Proton versions, but just with switching to them.

The previously mentioned workaround is still good, though. I can get the game to run by forcing the use of Proton 5.0-10 via the game's properties in Steam after `stl` has done its WMP10 hack. In that case, the `launchSteamGame` logging is as follows:
```
Sat Nov  7 15:03:02 EST 2020 INFO - launchSteamGame - GAMESTARTCMD is '/home/michael/.steam/steam/steamapps/common/Proton 5.0/proton waitforexitandrun /home/michael/.steam/steam/steamapps/common/Giana Sisters Twisted Dreams/launcher/GSLauncher.exe'
Sat Nov  7 15:03:02 EST 2020 SKIP - launchSteamGame - USESLR is enabled, but RUNSLR is empty - skipping
Sat Nov  7 15:03:02 EST 2020 INFO - wantProton - Proton version 'proton-5.0-10b' required, but '' is set
Sat Nov  7 15:03:02 EST 2020 INFO - wantProton - Proton version 'proton-5.0-10b' does exist - switching
Sat Nov  7 15:03:02 EST 2020 INFO - switchProton - Switching used Proton and its Variables to version 'proton-5.0-10b'
Sat Nov  7 15:03:02 EST 2020 INFO - switchProton - requested Proton version 'proton-5.0-10b' found under '/home/michael/.steam/steam/steamapps/common/Proton 5.0/proton'
Sat Nov  7 15:03:02 EST 2020 INFO - setNewProtVars - Setting new Proton Variables based on '/home/michael/.steam/steam/steamapps/common/Proton 5.0/proton'
Sat Nov  7 15:03:02 EST 2020 INFO - setNewProtVars - set the wine binary for proton in path '/home/michael/.steam/steam/steamapps/common/Proton 5.0/proton' to '/home/michael/.steam/steam/steamapps/common/Proton 5.0/dist/bin/wine'
Sat Nov  7 15:03:03 EST 2020 INFO - setProtonPathVersion Setting the Proton Version for '/home/michael/.steam/steam/steamapps/common/Proton 5.0/proton' to 'proton-5.0-10b'
Sat Nov  7 15:03:03 EST 2020 INFO - launchSteamGame - Proton Override allowed, so checking if it needs updated
Sat Nov  7 15:03:03 EST 2020 INFO - launchSteamGame - command line proton '/home/michael/.steam/steam/steamapps/common/Proton 5.0/proton' is identical to 'RUNPROTON' (path for 'proton-5.0-10b') - nothing to change
Sat Nov  7 15:03:03 EST 2020 INFO - launchSteamGame - using ORIGINAL game start command: /home/michael/.steam/steam/steamapps/common/Proton 5.0/proton waitforexitandrun /home/michael/.steam/steam/steamapps/common/Giana Sisters Twisted Dreams/launcher/GSLauncher.exe
Sat Nov  7 15:03:03 EST 2020 INFO - setProtonPathVersion Setting the Proton Version for '/home/michael/.steam/steam/steamapps/common/Proton 5.0/proton' to 'proton-5.0-10b'
Sat Nov  7 15:03:03 EST 2020 INFO - launchSteamGame - starting 'Giana Sisters Twisted Dreams (223220)' with Proton: 'proton-5.0-10b'
Sat Nov  7 15:03:03 EST 2020 INFO - launchSteamGame - full start command is '/home/michael/.steam/steam/steamapps/common/Proton 5.0/proton waitforexitandrun /home/michael/.steam/steam/steamapps/common/Giana Sisters Twisted Dreams/launcher/GSLauncher.exe'
```
There is a difference between the launch commands used in these two cases, namely that Steam Linux Runtime is used when I let STL switch Proton versions. So maybe I just need to disable that option. I'll try that next.

**Comment by michaelneverwins** [comment]
I disabled Steam Linux Runtime for the game via the Settings button on the STL window that pops up while the game is launching, and that seems to have fixed the problem. With that option unchecked, and the `stl` script modified as shown in my previous comment, I was able to launch the game successfully without forcing the correct Proton version via the Steam settings.

The `launchSteamGame` logging shows no Steam Linux Runtime was used:
```
Sat Nov  7 15:34:23 EST 2020 INFO - launchSteamGame - GAMESTARTCMD is '/home/michael/.steam/steam/steamapps/common/Proton 5.13/proton waitforexitandrun /home/michael/.steam/steam/steamapps/common/Giana Sisters Twisted Dreams/launcher/GSLauncher.exe'
Sat Nov  7 15:34:23 EST 2020 SKIP - launchSteamGame - USESLR is disabled, so skipping 'SteamLinuxRuntime' found in the commandline: '/home/michael/.steam/steam/steamapps/common/SteamLinuxRuntime_soldier/_v2-entry-point --deploy=soldier --suite=soldier --verb=waitforexitandrun --'
Sat Nov  7 15:34:23 EST 2020 INFO - wantProton - Proton version 'proton-5.0-10b' required, but 'proton-5.13-1b' is set
Sat Nov  7 15:34:23 EST 2020 INFO - wantProton - Proton version 'proton-5.0-10b' does exist - switching
Sat Nov  7 15:34:24 EST 2020 INFO - switchProton - Switching used Proton and its Variables to version 'proton-5.0-10b'
Sat Nov  7 15:34:24 EST 2020 INFO - switchProton - requested Proton version 'proton-5.0-10b' found under '/home/michael/.steam/steam/steamapps/common/Proton 5.0/proton'
Sat Nov  7 15:34:24 EST 2020 INFO - setNewProtVars - Setting new Proton Variables based on '/home/michael/.steam/steam/steamapps/common/Proton 5.0/proton'
Sat Nov  7 15:34:24 EST 2020 INFO - setNewProtVars - set the wine binary for proton in path '/home/michael/.steam/steam/steamapps/common/Proton 5.0/proton' to '/home/michael/.steam/steam/steamapps/common/Proton 5.0/dist/bin/wine'
Sat Nov  7 15:34:24 EST 2020 INFO - setProtonPathVersion Setting the Proton Version for '/home/michael/.steam/steam/steamapps/common/Proton 5.0/proton' to 'proton-5.0-10b'
Sat Nov  7 15:34:24 EST 2020 INFO - launchSteamGame - Proton Override allowed, so checking if it needs updated
Sat Nov  7 15:34:24 EST 2020 INFO - launchSteamGame - ORIGINAL game start command is: /home/michael/.steam/steam/steamapps/common/Proton 5.13/proton waitforexitandrun /home/michael/.steam/steam/steamapps/common/Giana Sisters Twisted Dreams/launcher/GSLauncher.exe
Sat Nov  7 15:34:24 EST 2020 INFO - launchSteamGame - overriding 'proton-5.13-1b' (='/home/michael/.steam/steam/steamapps/common/Proton 5.13/proton') from command line with 'proton-5.0-10b' (='/home/michael/.steam/steam/steamapps/common/Proton 5.0/proton')
Sat Nov  7 15:34:24 EST 2020 INFO - launchSteamGame - game was started with 'stl' as Launch option, so replacing '/home/michael/.steam/steam/steamapps/common/Proton 5.13/proton' with '/home/michael/.steam/steam/steamapps/common/Proton 5.0/proton' in '/home/michael/.steam/steam/steamapps/common/Proton 5.13/proton waitforexitandrun /home/michael/.steam/steam/steamapps/common/Giana Sisters Twisted Dreams/launcher/GSLauncher.exe'
Sat Nov  7 15:34:24 EST 2020 INFO - launchSteamGame - new launch command is '/home/michael/.steam/steam/steamapps/common/Proton 5.0/proton waitforexitandrun /home/michael/.steam/steam/steamapps/common/Giana Sisters Twisted Dreams/launcher/GSLauncher.exe'
Sat Nov  7 15:34:24 EST 2020 INFO - launchSteamGame - UPDATED game start command is: /home/michael/.steam/steam/steamapps/common/Proton 5.0/proton waitforexitandrun /home/michael/.steam/steam/steamapps/common/Giana Sisters Twisted Dreams/launcher/GSLauncher.exe
Sat Nov  7 15:34:24 EST 2020 INFO - setProtonPathVersion Setting the Proton Version for '/home/michael/.steam/steam/steamapps/common/Proton 5.0/proton' to 'proton-5.0-10b'
Sat Nov  7 15:34:24 EST 2020 INFO - launchSteamGame - starting 'Giana Sisters Twisted Dreams (223220)' with Proton: 'proton-5.0-10b'
Sat Nov  7 15:34:24 EST 2020 INFO - launchSteamGame - full start command is '/home/michael/.steam/steam/steamapps/common/Proton 5.0/proton waitforexitandrun /home/michael/.steam/steam/steamapps/common/Giana Sisters Twisted Dreams/launcher/GSLauncher.exe'
```

I wonder if anyone else has trouble running this game with Steam Linux Runtime. The option was enabled in STL by default, and I just hadn't noticed it was breaking the game until now. But anyone trying to reproduce that result would need to make sure Steam is configured to run the game with the "wrong" Proton version. STL didn't use Steam Linux Runtime when Steam itself was already configured to run the game with Proton 5.0-10; the option was used only when STL had to switch me from Proton 5.13-1 to Proton 5.0-10.

**Comment by frostworx** [comment]
Thanks for the extensive testing! Recently pretty much changed in the code (without much testing) so chances were high, that something is breaking in between.
I didn't have my first coffee yet, but I'm sure we'll find the problem. :)

**Comment by frostworx** [comment]
Just started to read your comments backwards to see what the latest state is.
The "Steam Linux Runtime (SLR)" which you see enabled is not enabled by stl, but by Proton and therefore is "enabled" by default.
It already caused massive problems with proton 5.13 and if unchecking it in stl helps to get the game working although proton-5.0.x is used that means, that SLR was also added to the recently released [proton-5.0-10b](https://github.com/ValveSoftware/Proton/releases/tag/proton-5.0-10b), because until then only 5.13 started SLR soldier that way.

**Comment by frostworx** [comment]
Good find @ https://github.com/frostworx/steamtinkerlaunch/issues/123#issuecomment-723489060
This is clearly a bug - commiting the fix now, thanks!

**Comment by frostworx** [comment]
Afaik the "blank space" you mention can be found in every dropdown menu where a value is preconfigured,
never checked if this is a glitch in yad or on my end.

**Comment by frostworx** [comment]
Btw Proton only calls the "SteamLinux Runtime (SLR)" with the method introduced with Proton 5.13, when not using a compatibility tool. So if you use steamtinkerlaunch as compatibility tool instead of launch parameter it isn't loaded as well and Giana should work if SLR really causes the issue here.
Will test myself now

**Comment by frostworx** [comment]
Yes, I can reproduce that "disabling SLR" (by using steamtinkerlaunch as compatibiliy tool instead of launch parameter) is required *now* to get Giana working.
I won't disable SLR in the tweak automatically though:
https://github.com/ValveSoftware/Proton/issues/4278#issuecomment-710429664
https://github.com/ValveSoftware/Proton/issues/4278#issuecomment-710463026

So if this is the only issue left why Giana currently/no longer working with stl, I'm afraid this is not my bug anymore.
_(or did I miss something)_

**Comment by frostworx** [comment]
Tried to sum up the problem in [this sticky](https://github.com/frostworx/steamtinkerlaunch/issues/140)

**Comment by michaelneverwins** [comment]
I reverted my change to the script, pulled the latest changes from the master branch (including 39637dc15f3d0de8e00b204975ddf2517e95027b), and tested again with
```bash
rm -rf ~/.config/stl
rm -rf /dev/shm/stl
stl settings
```
and the script was able to find the official Proton versions including 5.0-10b.

As for Steam Linux Runtime, I'm still not certain if I understand how it works or how STL interacts with it, but I've confirmed that I'm able to run the game by using SteamTinkerLaunch as a Steam Play compatibility tool instead of a launch option. (I ran `stl compat add`, deleted `~/.config/stl` and `/dev/shm/stl` as well as `.steam/steam/steamapps/compatdata/223220`, restarted Steam, selected SteamTinkerLaunch as the game's compatibility tool and removed `stl %command%` from the launch options, and started the game. The WMP10 fix was applied, the correct Proton version was detected and used, and the game worked.) This works even with Proton 5.13-1 selected in my global Steam settings, so I can see that SteamTinkerLaunch as a compatibility tool is automatically choosing Proton 5.0-10 based on the game's tweak file and running it successfully without Steam Linux Runtime.

I noticed this line in `~/.config/stl/logs/223220.log`:
```
Sun Nov  8 11:10:24 EST 2020 SKIP - launchSteamGame - USESLR is enabled, but RUNSLR is empty - skipping
```
I'm guessing the `USESLR` variable is controlled by the "Use Steam Linux Runtime" checkbox in the settings which is enabled by default, whereas `RUNSLR` is determined by whether Steam is trying to run the game with SLR, which in turn depends on whether a specific Proton version is selected as the game's compatibility tool. Is that accurate? If so — and if Steam uses SLR by default only with the latest Proton — then it would explain why SLR broke the game when I ran it with Steam configured to use Proton 5.13-1, whereas SLR was *not* used when I selected Proton 5.0-10 via Steam. In the logs posted earlier, it looks like STL saw SLR in the launch command which Steam built for Proton 5.13-1, and kept SLR even while switching to Proton 5.0-10, which caused the game to crash. I guess the lesson learned here is that using STL as a compatibility tool is a bit more foolproof than using it as a launch option, and in the latter case, one has to pay a little more attention to what settings are being used.

Anyway, the problem with detecting Proton versions appears to be solved, so I'm closing the issue again. Thanks @frostworx!

**Comment by frostworx** [comment]
Nice, thanks for testing,  @michaelneverwins!

I try to explain it differently to make it better understandable:

Totally unrelated from steamtinkerlaunch the SLR is only part of the game launch parameters/used when the game is not started using any custom "compatibility tool".
As steamtinkerlaunch is _(afaik)_ the only tool which can be started as launch option an as compatibility tool it is easy to see the difference when using it with one and then the other way.

You can easily check the output `/dev/shm-cmdline.txt` of the one liner I added earlier today
_(also appended the change in [this sticky](https://github.com/frostworx/steamtinkerlaunch/issues/140)_

So disabling SLR in steamtinkerlaunch only has any effect when steamtinkerlaunch is used as launch option _(because it is disabled anyway when used as compatibility tool)._

`RUNSLR` is the Steam/Proton internal SLR launch parameter part, which is empty when using a compatibility tool.
That is what above message 
`Sun Nov  8 11:10:24 EST 2020 SKIP - launchSteamGame - USESLR is enabled, but RUNSLR is empty - skipping`
is trying to say.

When you use no compatibility tool and use `stl %command%` instead, the `RUNSLR` part is active _(check the one-liner)_
and can be disabled with unchecking `USESLR` in steamtinkerlauch `(this does nothing else than stripping out the `RUNSLR` part from the command line before starting the game)`

Feel free to ask if something is still unclear. I don't think that there are many users who know this, but imho it is good to be known because it helps to solve problems.
